# Лабораторная работа 1. Язык Python, основные библиотеки для анализа данных.

ФИО: Кибитова Валерия Николаевна

Группа: 517

In [1]:
import numpy as np
import pandas as pd

## Часть 1: Python

#### 1. Напишите код, который сформирует список всех натуральных чисел, не превосходящих 100, которые делятся на 7 и не делятся на 5. Нельзя использовать циклы — выполните это задание с помощью list comprehension.

In [4]:
l = [x for x in xrange(1, 101) if (x % 7 == 0) and (x % 5 != 0)]
l

[7, 14, 21, 28, 42, 49, 56, 63, 77, 84, 91, 98]

#### 2. Напишите функцию `sort_tokens(s)`, которая принимает на вход строку, разделяет ее на токены по запятым, сортирует токены с помощью функции `sorted()` и снова соединяет их в одну строку через запятые. Например, для строки `u"ночь,фонарь,улица"` функция должна выдать `u"ночь,улица,фонарь"`. Придумайте несколько входных строк и протестируйте функцию на них.

Обратите внимание, что склеить строки можно с помощью конструкции `'x'.join(...)`. Например:

In [2]:
'! '.join(['i', 'am', 'very', 'excited', ':-)'])

'i! am! very! excited! :-)'

In [9]:
def sort_tokens(s):
    s = s.split(',')
    s = sorted(s)
    s = ','.join(s)
    return s
print sort_tokens("python,c++,haskell,java,ruby")
print sort_tokens("cat,dog,hamster,fox")

c++,haskell,java,python,ruby
cat,dog,fox,hamster


#### 3. Напишите функцию `remove_duplicates(a)`, которая принимает на вход список и возвращает его же, но без дубликатов и в отсортированном виде. Например, для списка `[u"ночь", u"улица", u"фонарь", u"аптека", u"аптека", u"улица", u"фонарь"]` результат должен быть `[u"аптека", u"ночь", u"улица", u"фонарь", ]`.

Подсказка: обратите внимание на структуру `set`!

In [11]:
def remove_duplicates(a):
    return sorted(set(a))
remove_duplicates(["a", "bb", "aa", "a", "z", "z", "d"])

['a', 'aa', 'bb', 'd', 'z']

#### 4. Напишите функцию `word_counts(texts)`, которая принимает на вход список строк, и печатает пары (слово, число вхождений). Считайте, что слова в текстах могут быть разделены только пробелами. Например, для входа `[u'ночь улица фонарь аптека', u'аптека улица фонарь']` должно быть напечатано
>аптека 2

>улица 2

>фонарь 2

>ночь 1

Подсказки:

Чтобы разбить каждый из текстов на слова, воспользуйтесь функцией `map`, которая применяет заданную функцию к каждому элементу списка. Например, если бы нужно было каждый из текстов привести к нижнему регистру, то команда выглядела бы так:

In [3]:
map(lambda x: x.lower(), ["abCDefg", "HIjkL"])

['abcdefg', 'hijkl']

После первого шага вы получите список, состоящий из списков слов. Чтобы объединить их в один список слов, воспользуйтесь функцией `reduce`. Например, просуммировать числа из списка с ее помощью можно так:



In [4]:
reduce(lambda x, y: x + y, [1, 2, 3, 4, 5])

15

Для подсчета количества вхождений слов воспользуйтесь структурой `defaultdict` из модуля `collections`. В отличие от обычного `dict`, она не генерирует ошибку при обращении к несуществущему ключу, а создает новую запись, инициализируемую нулевым значением. Например:

In [18]:
from collections import defaultdict
d = defaultdict(int)
print d
d['abc'] += 1
print d
d['xxx']
print d

defaultdict(<type 'int'>, {})
defaultdict(<type 'int'>, {'abc': 1})
defaultdict(<type 'int'>, {'xxx': 0, 'abc': 1})


In [22]:
def word_counts(texts):
    texts = map(lambda x: x.split(' '), texts)
    texts = reduce(lambda x, y: x + y, texts)
    words = defaultdict(int)
    for i in texts:
        words[i] += 1
    for key in words:
        print '{0} {1}'.format(key, words[key])
    
word_counts(["a bc", "d f", "ee f d a", "a a a", "z"]) 

a 5
d 2
f 2
ee 1
bc 1
z 1


#### 5. (бонусное задание) Скачайте английский перевод "Преступления и наказания". Найдите 20 самых популярных слов и 20 слов, которые встречаются только один раз.

Приведите текст к нижнему регистру. Разделите текст на слова, считая разделителями все, кроме букв. Обратите внимание на функцию `split` из модуля `re`.

Не забудьте удалить пустые слова:
```
words = filter(lambda x: len(x) > 0, words)
```

In [25]:
from urllib import urlopen
url = "http://www.gutenberg.org/files/2554/2554.txt"
response = urlopen(url)
raw_text = response.read().decode('utf8')

In [26]:
raw_text[1000:1100]

u'r five\r\nchildren in only two rooms. The father and mother spent their evenings\r\nin reading aloud to '

In [46]:
import re
from collections import Counter
words = re.split("[^A-Za-z]+", raw_text.lower())
words = filter(lambda x: len(x) > 0, words)
counter = Counter(words)
print [str(x[0]) for x in counter.most_common(20)]
print [str(x[0]) for x in counter.most_common()[-20:]]

['the', 'and', 'to', 'he', 'a', 'i', 'you', 'of', 'it', 'that', 'in', 'was', 'his', 'at', 'her', 'but', 'not', 's', 'with', 'she']
['lawful', 'fawning', 'sabre', 'noxious', 'faithfully', 'uncontrollable', 'funds', 'deuce', 'tactics', 'rigamarole', 'venetian', 'rickety', 'unfolding', 'gnawing', 'bombarding', 'transfigured', 'june', 'upwards', 'sash', 'bereaved']


20 самых популярных слов: 'the', 'and', 'to', 'he', 'a', 'i', 'you', 'of', 'it', 'that', 'in', 'was', 'his', 'at', 'her', 'but', 'not', 's', 'with', 'she'

20 слов, которые встречаются только 1 раз: 'lawful', 'fawning', 'sabre', 'noxious', 'faithfully', 'uncontrollable', 'funds', 'deuce', 'tactics', 'rigamarole', 'venetian', 'rickety', 'unfolding', 'gnawing', 'bombarding', 'transfigured', 'june', 'upwards', 'sash', 'bereaved'

## Часть 2: NumPy и SciPy

#### 1. Повторите 1000 раз следующий эксперимент: сгенерируйте две матрицы размера $10 \times 10$ из стандартного нормального распределения, перемножьте их (как матрицы) и найдите максимальный элемент. Какое среднее значение по экспериментам у максимальных элементов? 95-процентная квантиль?

Для вычисления квантили воспользуйтесь функцией `np.percentile`

In [73]:
from numpy.random import randn
import numpy as np
max_list = []
for i in xrange(1000):
    m_1 =  randn(10, 10)
    m_2 = randn(10, 10)
    m = m_1.dot(m_2)
    max_list.append(m.max())
print np.mean(max_list)
print np.percentile(max_list, 95)

8.38557183558
11.6800640541


Среднее значение по экспериментам у максимальных элементов: 8.38557183558

95-процентная квантиль: 11.6800640541

#### 2. Сгенерируйте 1000 чисел из распределения Пуассона с параметром $\lambda = 5$. Сформируйте массив, в котором в i-м элементе будет записано количество сгенерированных чисел, равных i.

Обратите внимание на функцию `np.bincount`

In [70]:
from numpy.random import poisson
ar = poisson(5, size=1000)
ar = np.bincount(ar)
print ar

[  6  41  88 137 187 154 139 115  76  31  14   8   1   2   1]


#### 3. Реализуйте функцию, которая принимает на вход numpy-массив целых чисел `a`, и генерирует массив, в котором число `i` встречается `a[i]` раз. Протестируйте на результате из прошлого задания.

Обратите внимание на функцию `np.repeat`

In [74]:
def generate_array(ar):
    return np.repeat(range(len(ar)),ar)
print generate_array(ar)

[ 0  0  0  0  0  0  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  2  2  2
  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
  2  2  2  2  2  2  2  2  2  2  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  4  4  4
  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4
  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4
  4  4  4  4

#### 4. Напишите функцию `scale(X)`, которая принимает на вход матрицу и масштабирует каждый ее столбец (вычитает среднее и делит на стандартное отклонение). Убедитесь, что в функции не будет происходить деления на ноль. Протестируйте на каких-нибудь данных.

In [91]:
def scale(X):
    std = np.std(X, axis=0)
    std = [1 if x==0 else x for x in std]
    mean = np.mean(X, axis = 0)
    X = (X - mean) / std
    return X
X = np.array([[2,2,2],[2, 5,19]])
print scale(X)
print scale(m_1)

[[ 0. -1. -1.]
 [ 0.  1.  1.]]
[[-0.42088573  0.02000709  0.48121942 -1.32604851  0.48230758  0.67583164
  -1.3487649  -0.42852779  0.13193867  1.0871897 ]
 [ 0.1173854  -0.24069303 -0.6161566   0.50853403  0.75044555 -0.24077013
  -0.4402397  -0.81538703  2.43334288 -1.45738846]
 [-0.15806883 -0.85276809  0.47238362  0.4348827  -0.04363508 -0.537303
   0.55241296 -0.28000893 -0.40405193 -0.31204093]
 [-1.48743691  1.75730617  0.03976111 -1.0095925  -0.9667377   0.9848162
   1.76280785  1.00322147 -0.07680156  0.62766903]
 [-0.97056648 -0.57973047 -0.68452654  0.26237922 -0.93952726 -1.08835921
   0.5908394  -0.52426422 -0.14825386 -0.45454958]
 [ 0.27040039  0.83777953  1.12751191  0.25656971  0.11943351 -0.26415893
  -1.62433867  0.89080724 -0.98699912  0.34763326]
 [ 1.92013551  0.27214432  1.08113844 -0.82911557 -1.97581779 -0.38179952
  -0.02348688 -0.64834393 -0.04575588  1.19071965]
 [ 1.51712556  1.07338556 -1.87481328 -0.92957243  0.37617814 -0.60812704
   0.48822604 -1.770691

#### 5. Сгенерируйте матрицу с элементами из нормального распределения $\mathcal{N}(10, 1)$. Найдите ее:
- определитель
- след
- наименьший и наибольший элементы
- спектральную норму
- норму Фробениуса
- собственные числа
- обратную матрицу

In [112]:
from  numpy.linalg import det
from  numpy import trace
from numpy.linalg import norm
from numpy.linalg import inv
from numpy.linalg import eig
m =  randn(5, 5) + 10
print m
det_m = det(m)
print det_m
trace_m = trace(m)
print trace_m
max_m = np.max(m)
print max_m
min_m = np.min(m)
print min_m
s_norm = norm(m, ord=2)
print s_norm
f_norm = norm(m, ord='fro')
print f_norm
eig, eig_v = eig(m)
print eig
inv_m = inv(m)
print inv_m


[[ 11.53100341  10.48515278  10.40831633  11.00404944  10.65440731]
 [  9.09081409   9.56133319  10.56164335  10.42636519   8.55456311]
 [  8.95207569  11.25682857  10.42235979   9.17199581   9.22039642]
 [  9.96306731   8.4509371    9.45730517   9.71663174  10.74356422]
 [  9.29556908   8.11008263  10.42842563  11.07466223   9.7319856 ]]
35.3684447712
50.9633137324
11.5310034135
8.11008262607
49.724069738
49.8619612293
[ 49.64013108  -1.7279682    2.16704049  -0.17898197   1.06309233]
[[-0.37669181  2.29314712 -1.19120027  1.42710459 -2.05017719]
 [ 0.76593907 -1.68678121  0.90361035 -1.28242233  1.20378215]
 [-1.74006043  3.55948003 -1.29785971  2.44097583 -2.68891515]
 [ 1.3167284  -2.85076438  1.02396886 -2.31103061  2.64544062]
 [ 0.08770093 -1.35477931  0.61026128 -0.28019459  0.92874997]]


матрица = 
          
          [ [ 11.53100341  10.48515278  10.40831633  11.00404944  10.65440731]

          [  9.09081409   9.56133319  10.56164335  10.42636519   8.55456311]
          
          [  8.95207569  11.25682857  10.42235979   9.17199581   9.22039642]
          
          [  9.96306731   8.4509371    9.45730517   9.71663174  10.74356422]
          
          [  9.29556908   8.11008263  10.42842563  11.07466223   9.7319856 ]]

определитель = 35.3684447712

след = 50.9633137324

наименьший элемент = 8.11008262607

наибольший элемент = 11.5310034135

спектральная норма = 49.724069738

норма фробениуса = 49.8619612293

собственные числа = [ 49.64013108  -1.7279682    2.16704049  -0.17898197   1.06309233]

обратная матрица = 

                    [[-0.37669181  2.29314712 -1.19120027  1.42710459 -2.05017719]
                    
                    [ 0.76593907 -1.68678121  0.90361035 -1.28242233  1.20378215]
                    
                    [-1.74006043  3.55948003 -1.29785971  2.44097583 -2.68891515]
                    
                    [ 1.3167284  -2.85076438  1.02396886 -2.31103061  2.64544062]
                    
                    [ 0.08770093 -1.35477931  0.61026128 -0.28019459  0.92874997]]

## Часть 3: Pandas

#### Ответьте на вопросы о данных по авиарейсам в США за январь-апрель 2008 года.

Данные: http://stat-computing.org/dataexpo/2009/2008.csv.bz2
(обратите внимание, что распаковывать этот файл не обязательно — функция `pandas.read_csv` умеет читать из архивов автоматически)

Описание: http://stat-computing.org/dataexpo/2009/the-data.html

1. Какая из причин отмены рейса (`CancellationCode`) была самой частой? (расшифровки кодов можно найти в описании данных)
2. Найдите среднее, минимальное и максимальное расстояние, пройденное самолетом.
3. Не выглядит ли подозрительным минимальное пройденное расстояние? В какие дни и на каких рейсах оно было? Какое расстояние было пройдено этими же рейсами в другие дни?
4. Из какого аэропорта было произведено больше всего вылетов? В каком городе он находится?
5. Найдите для каждого аэропорта среднее время полета (`AirTime`) по всем вылетевшим из него рейсам. Какой аэропорт имеет наибольшее значение этого показателя?
6. Найдите аэропорт, у которого наибольшая доля задержанных (`DepDelay > 0`) рейсов. Исключите при этом из рассмотрения аэропорты, из которых было отправлено меньше 1000 рейсов (используйте функцию `filter` после `groupby`).

In [3]:
import pandas as pd
df = pd.read_csv("2008.csv")
df.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2008      1           3          4     2003        1955     2211   
1  2008      1           3          4      754         735     1002   
2  2008      1           3          4      628         620      804   
3  2008      1           3          4      926         930     1054   
4  2008      1           3          4     1829        1755     1959   

   CRSArrTime UniqueCarrier  FlightNum        ...         TaxiIn  TaxiOut  \
0        2225            WN        335        ...              4        8   
1        1000            WN       3231        ...              5       10   
2         750            WN        448        ...              3       17   
3        1100            WN       1746        ...              3        7   
4        1925            WN       3920        ...              3       10   

   Cancelled  CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0          0               NaN         0           NaN          NaN      NaN   
1          0               NaN         0           NaN          NaN      NaN   
2          0               NaN         0           NaN          NaN      NaN   
3          0               NaN         0           NaN          NaN      NaN   
4          0               NaN         0             2            0        0   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
3            NaN                NaN  
4              0                 32  

[5 rows x 29 columns]

In [4]:
df['CancellationCode'].value_counts().idxmax()

'B'

1) Самая частая причина отмены рейсов погода("weather")

In [5]:
df["Distance"].mean()

726.38702942539283

In [6]:
df["Distance"].max()

4962

In [7]:
df["Distance"].min()

11

2) Среднее расстояние пройденное самолетом 726.38702942539283

  Максимальное расстояние пройденное самолетом 4962
  
  Минимальное расстояние пройденное самолетом 11

In [8]:
df[["Year", "Month", "DayofMonth", "FlightNum", "TailNum"]][df["Distance"] == 11]

Year  Month  DayofMonth  FlightNum TailNum
2547298  2008      5          15       4988  N806CA
4392215  2008      8          10       5572  N819CA

In [9]:
df["Distance"][(df["FlightNum"]==4988) & (df["TailNum"]=="N806CA") & ((df["Year"] != 2008) | (df["Month"] != 5) | (df["DayofMonth"]!=15))].unique()

array([834])

In [10]:
df["Distance"][(df["FlightNum"]==5572) & (df["TailNum"]=="N819CA") & ((df["Year"] != 2008)| (df["Month"] != 8) | (df["DayofMonth"]!=10))].unique()

array([544])

3) Выглядит. Оно было 15 мая 2008 на рейсе 4988 N806CA и 10 августа 2008 на рейсе 5572 N819CA.

В другие дни рейс 4988 пролетал одно из этих значений: 834 мили. 

В другие дни рейс 5572 N819CA пролетал 544 мили. 

In [12]:
airport_id = df['Origin'][df['Cancelled'] == 0].value_counts().idxmax()
df_a = pd.read_csv("airports.csv")
df_a[["city", "airport"]][df_a["iata"] == airport_id]

city                            airport
880  Atlanta  William B Hartsfield-Atlanta Intl

4)Больше всего вылетов было произведено из аэропорта William B Hartsfield-Atlanta Intl, который находится в Atlanta	

In [13]:
df[df["Cancelled"] == 0].groupby('Origin').AirTime.mean()

Origin
ABE        89.057518
ABI        36.789977
ABQ        91.657166
ABY        35.379759
ACK        50.692124
ACT        29.875977
ACV        58.850658
ACY       108.848214
ADK       151.423913
ADQ        41.987934
AEX        68.543556
AGS        36.933814
AKN        37.321429
ALB       101.860527
ALO        36.373134
...
TUP       42.900000
TUS       92.801864
TVC       49.328770
TWF       41.316563
TXK       43.796312
TYR       30.372715
TYS       81.030894
VLD       45.158351
VPS       75.231116
WRG       16.741606
WYS       62.961977
XNA       84.253180
YAK       36.372159
YKM       80.697329
YUM       48.876952
Name: AirTime, Length: 303, dtype: float64

In [14]:
df[df["Cancelled"] == 0].groupby('Origin').AirTime.mean().idxmax()

'SJU'

5) Наиболбшее среднее время полета имеет аэропорт 'SJU'

In [15]:
z = lambda x: len(x[x > 0]) * 1.0 / len(x)
df.groupby('Origin').filter(lambda x:len(x) > 1000).groupby('Origin').DepDelay.agg(z).idxmax()


'DAL'

6)Наибольшая доля задержанных рейсов у аэропорта 'DAL'